internal/ratelimit/burst_test.go

```diff
@@ -0,0 +1,187 @@
+package ratelimit
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/supabase/auth/internal/conf"
+)
+
+func Example_newBurstLimiter() {
+	now, _ := time.Parse(time.RFC3339, "2024-09-24T10:00:00.00Z")
+	{
+		cfg := conf.Rate{Events: 10, OverTime: time.Second * 20}
+		rl := NewBurstLimiter(cfg)
+		cur := now
+		for i := 0; i < 20; i++ {
+			allowed := rl.AllowAt(cur)
+			fmt.Printf("%-5v @ %v\n", allowed, cur)
+			cur = cur.Add(time.Second * 5)
+		}
+	}
+
+	// Output:
+	// true  @ 2024-09-24 10:00:00 +0000 UTC
+	// true  @ 2024-09-24 10:00:05 +0000 UTC
+	// true  @ 2024-09-24 10:00:10 +0000 UTC
+	// true  @ 2024-09-24 10:00:15 +0000 UTC
+	// true  @ 2024-09-24 10:00:20 +0000 UTC
+	// true  @ 2024-09-24 10:00:25 +0000 UTC
+	// true  @ 2024-09-24 10:00:30 +0000 UTC
+	// true  @ 2024-09-24 10:00:35 +0000 UTC
+	// true  @ 2024-09-24 10:00:40 +0000 UTC
+	// true  @ 2024-09-24 10:00:45 +0000 UTC
+	// true  @ 2024-09-24 10:00:50 +0000 UTC
+	// true  @ 2024-09-24 10:00:55 +0000 UTC
+	// true  @ 2024-09-24 10:01:00 +0000 UTC
+	// false @ 2024-09-24 10:01:05 +0000 UTC
+	// false @ 2024-09-24 10:01:10 +0000 UTC
+	// false @ 2024-09-24 10:01:15 +0000 UTC
+	// true  @ 2024-09-24 10:01:20 +0000 UTC
+	// false @ 2024-09-24 10:01:25 +0000 UTC
+	// false @ 2024-09-24 10:01:30 +0000 UTC
+	// false @ 2024-09-24 10:01:35 +0000 UTC
+}
+
+func TestBurstLimiter(t *testing.T) {
+	t.Run("Allow", func(t *testing.T) {
+		for i := 1; i < 10; i++ {
+			cfg := conf.Rate{Events: float64(i), OverTime: time.Hour}
+			rl := NewBurstLimiter(cfg)
+			for y := i; y > 0; y-- {
+				if exp, got := true, rl.Allow(); exp != got {
+					t.Fatalf("exp Allow() to be %v; got %v", exp, got)
+				}
+			}
+			if exp, got := false, rl.Allow(); exp != got {
+				t.Fatalf("exp Allow() to be %v; got %v", exp, got)
+			}
+		}
+	})
+
+	t.Run("AllowAt", func(t *testing.T) {
+		now, _ := time.Parse(time.RFC3339, "2024-09-24T10:00:00.00Z")
+
+		type event struct {
+			ok bool
+			at time.Time
+			r  int
```

Comment:
what does `r` represent here? reading the code makes me think number of requests? would be good to leave a comment or rename the field so the test is easier to follow

Reply:
r for requests, rounds, repeats, could have been `i` instead, just a loop for the number of times to execute.

```Go
			for _, evt := range tc.evts {
				for i := 0; i <= evt.r; i++ {
					if exp, got := evt.ok, rl.AllowAt(evt.at); exp != got {
						t.Fatalf("exp AllowAt(%v) to be %v; got %v", evt.at, exp, got)
					}
				}
			}
```